sw/device/lib/dif/dif_resetmgr.h

```diff
@@ -0,0 +1,187 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_RESETMGR_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_RESETMGR_H_
+
+/**
+ * @file
+ * @brief <a href="">Reset Manager</a> Device Interface Functions
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Reset Manager generic enable/disable enumeration.
+ *
+ * Enumeration used to enable/disable bits, flags, ...
+ */
+typedef enum dif_resetmgr_enable {
+  kDifResetmgrDisable = 0, /**< disable. */
+  kDifResetmgrEnable,      /**< enable. */
+} dif_resetmgr_enable_t;
+
+/**
+ * Reset Manager peripheral software reset control.
+ */
+typedef enum dif_resetmgr_peripheral_reset {
+  /**
+   * Simple peripheral reset (release straight away).
+   */
+  kDifResetmgrPeripheralReset = 0,
+  kDifResetmgrPeripheralResetHold,    /**< Hold peripheral in reset. */
+  kDifResetmgrPeripheralResetRelease, /**< Release peripheral from reset. */
+} dif_resetmgr_peripheral_reset_t;
+
+/**
+ * Reset Manager software reset available peripherals.
+ */
+typedef enum dif_resetmgr_peripherals {
+  kDifResetmgrUsb = 0,
+  kDifResetmgrSpi,
+} dif_resetmgr_peripheral_t;
+
+/**
+ * Reset Manager all possible reset causes enumeration.
+ *
+ * TODO: no documentation currently available, would be useful
+ *       for writing a more detailed description for the individual causes.
+ */
+typedef enum dif_resetmgr_cause {
+  kDifResetmgrCausePowerUp = 0,
+  kDifResetmgrCauseLowPowerExit,
+  kDifResetmgrCauseWatchdog,
+  kDifResetmgrCauseSecurityEscalation,
+  kDifResetmgrCauseNonDebugModuleRequest,
+  kDifResetmgrCauseLast = kDifResetmgrCauseNonDebugModuleRequest,
+} dif_resetmgr_cause_t;
+
+/**
+ * Reset Manager reset causes information.
+ */
+typedef struct dif_resetmgr_causes {
+  bool cause[kDifResetmgrCauseLast + 1];
```

Comment:
I'm not sold on the ergonomics of `bool[N]` to represent N different reset causes. If the total number of resets will always be strictly less than 32 then let's encode these as one-hot enumerations. I know the difference is between:

```c
if (cause.cause[kDifResetmgrCausePowerUp]) {
```

And

```c
if (cause & kDifResetmgrCausePowerUp) {
```

on a normal single-condition check. But since the bits aren't mutually exclusive I anticipate that the access pattern will be "did we reset because of X and only X?" and that two `ands` and an `or` if it's packed into a u32. If it's an array we have to iterate and accumulate the conditions.

Does that seem reasonable?

Reply:
> If the total number of resets will always be strictly less than 32
That was one of the reasons I choose the array of bools, but I think you are right - I can't imagine there ever would even be 32 reset causes, nevermind more.

Indeed, this is something I also considered. I thought that array of bools was more suggestive to how it is supposed to be used. However, like you said - `cause.cause[kDifResetmgrCausePowerUp]` and `cause & kDifResetmgrCausePowerUp` are not too much different, and `cause & kDifResetmgrCausePowerUp` is more ergonomic of the two.

Reply:
> If it's an array we have to iterate and accumulate the conditions.

Maybe I am misunderstanding, but we could simply do:
```
if ((causes.cause[kDifResetmgrCausePowerUp] && causes.cause[kDifResetmgr*]) ||
      causes.cause[kDifResetmgr*]) {
```

Reply:
The condition I'm suggesting checking for is "our cause was _only_ PowerUp", for example...

```c
// Array of bools. Testing for _only_ kDifResetmgrCausePowerup.
if (!causes.cause[kDifResetmgrCausePowerUp]) {
  return false;
}
for (int i = 0; i < kDifResetmgrCauseCount; ++i) {
  if (i == kDifResetmgrCausePowerUp) {
    continue;
  }
  if (causes.cause[i] && i != (dif_cause_t)kDifResetmgrCausePowerUp) {
    return false;
  }
}
return true;

// Bitfields. Testing for _only_ kDifResetmgrCausePowerup.
return cause == kDifResetmgrCausePowerUp;
```

Reply:
I see, yeah, that makes sense.